### ISSUE TITLE
Exception in RegisterOrUpdateReminder (bug?) with SQL system store

### ISSUE THREAD

**Issue by kucheruk** [dotnet/orleans#637]
When using reminders with sql storage, call RegisterOrUpdateReminder for update reminder causes error:
Could not register reminder "..." to reminder table due to a race. Please try again later.
upd: exception thrown when calling RegisterOrUpdateReminder second time

LocalReminderService.cs:137
```C#
public async Task<IGrainReminder> RegisterOrUpdateReminder(GrainReference grainRef, string reminderName, TimeSpan dueTime, TimeSpan period)
{
    var entry = new ReminderEntry
    {
    ...
    };

    // entry.ETag is null 
    ...
    var newEtag = await reminderTable.UpsertRow(entry);

    //newEtag == null (!)
    if (newEtag != null) 
    {
    ....
    }

    var msg = string.Format("Could not register reminder {0} to reminder table due to a race. Please try again later.", entry);
    logger.Error(ErrorCode.RS_Register_TableError, msg);
    throw new ReminderException(msg);
}
```

```sql
UPDATE [OrleansRemindersTable]    
SET
    [StartTime]             = @startTime,
    [Period]                = @period,
    [GrainIdConsistentHash] = @grainIdConsistentHash
OUTPUT inserted.ETag
WHERE
    [ServiceId] = @serviceId AND @serviceId IS NOT NULL
    AND [GrainId] = @grainId AND @grainId IS NOT NULL
    AND [ReminderName] = @reminderName AND @reminderName IS NOT NULL
    AND ETag = @etag AND @etag IS NOT NULL; --   <---- ?
END
```
Looks like @etag here is always NULL.

**Comment by veikkoeeva** [comment]
@kucheruk That check means the return value from [SqlReminderTable.UpsertRow is null](https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/ReminderService/SqlReminderTable.cs#L79), the call and return value ends up at [OrleansRelationalExtensions.UpsertReminderRowAsync](https://github.com/dotnet/orleans/blob/master/src/Orleans/RelationalStorage/OrleansRelationalExtensions.cs#L395). The reason it can be null is that no row was inserted

As you may have noticed, the query is picked by key from ``[OrleansQuery]`` that is defined in in [CreateOrleansTables_SqlServer.sql](https://github.com/dotnet/orleans/blob/master/src/OrleansProviders/SQLServer/CreateOrleansTables_SqlServer.sql#L458). You can see in [here](https://github.com/dotnet/orleans/blob/master/src/OrleansProviders/SQLServer/CreateOrleansTables_SqlServer.sql#L461) the condition to either update to insert. You can just alter the row with a different condition to see if it works, e.g. checking just the ``ETag``.

I'll try to take a look at it today, but unfortunatelly I'm not 100% sure. I should be at home in few hours. If you have questions in the mean time, just ask here or in [Orleans Gitter](https://gitter.im/dotnet/orleans) and I'll try to answer. (Yep, I wrote that part and I may have made a last minute mistake.)

**Comment by gabikliot** [comment]
Ok, this is a bit of  a tricky code so let me explain what is going on here.
In general, usually, you do not want to use Upsert in a code that cares about concurrency control. You don't want to do just blind insert or update. Instead, you want to do TryInsert without etag and if it fails, TryUpadte with etag. That is what we do in the membership table case.
However, in the specific case of reminder operation `RegisterOrUpdateReminder` we don't care. The semantics of this operation are: I don't care if you had a previous instance, just overwrite it with the new data. All the data of the remainder (name, dueTiem, period) are overwritten and it is as if we created a new one. So in this case we do not need to actually TryUpdate. We just do blind overwrite.

If you look at Azure based impl. that is exactly what happens - we don't set etag on TableEntity and are calling `TableOperation.InsertOrReplace(data)` without etag.

So in the SQL case the right fix I think is actually not to check `AND ETag = @etag AND @etag IS NOT NULL ` but rather not to check for etag at all!

As a side note, the situation is actually a little bit more complicated. If you look carefully through the code, we actually do record the etag that was read, store it as part of our local reminder data and use it to compare sometimes. But this is different. In this case we use the etag  for our code, Orleans code, to catch the case that my silo had an older version of the reminder in memory and than the reminder in the table was updated and so my local copy of the reminder is different and I need to do something about it. So we use it to detect local inconsistency but we still never use to guarantee proper concurrency control in the table: we never use it when wring to the table.

**Comment by veikkoeeva** [comment]
@gabikliot This is a splendid, detailed explanation of the implementation rationale. It would be a shame to it get buried here. Can you propose a place to write it? If there isn't a better place and it is all right for you, I might just write it to the SQL script, submit it as a PR.

**Comment by gabikliot** [comment]
ye, I will write something, will try to... This is of course just a small part of the overall design. The overall design of the reminders subsystem, just only that one subsystem, has lots of complexities.
But maybe I will focus only on the parts that have external dependencies, like system store dependence (for MBR and Reminders) and explain what is required from the dependent components and the rationale. Will do.

**Comment by veikkoeeva** [comment]
@gabikliot But remember you are on holiday. This can wait.

**Comment by gabikliot** [comment]
Fixed by @kucheruk via #639.
